File: gtci/09-hashmaps/cpp/3-maximum-number-of-balloons.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <limits>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    int maxNumberOfBalloons(string text) {
        std::unordered_map<char, int> balloonFreqMap = {
            {'b', 1}, {'a', 1}, {'l', 2}, {'o', 2}, {'n', 1}};

        std::unordered_map<char, int> textFreqMap;
        for (char ch : text) textFreqMap[ch] += 1;

        int minCount = std::numeric_limits<int>::max();
        for (const auto& [ch, freq] : balloonFreqMap) {
            if (!textFreqMap.contains(ch)) return 0;

            int textMultiple = textFreqMap[ch] / balloonFreqMap[ch];
            minCount = std::min(minCount, textMultiple);
        }

        return minCount;
    }
};

int main() {
    Solution sol;

    assert(sol.maxNumberOfBalloons("nlaebolko") == 1);
    assert(sol.maxNumberOfBalloons("loonbalxballpoon") == 2);
    assert(sol.maxNumberOfBalloons("leetcode") == 0);
    assert(sol.maxNumberOfBalloons("balloonballoon") == 2);
    assert(sol.maxNumberOfBalloons("balloonballoooon") == 2);
    assert(sol.maxNumberOfBalloons("bbaall") == 0);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}

File: gtci/09-hashmaps/python3/3-maximum-number-of-balloons.py
class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        balloon_freq_map = {}
        for char in "balloon":
            if char not in balloon_freq_map:
                balloon_freq_map[char] = 0
            balloon_freq_map[char] += 1

        text_freq_map = {}
        for char in text:
            if char not in text_freq_map:
                text_freq_map[char] = 0
            text_freq_map[char] += 1

        min_count = float("inf")
        for char in balloon_freq_map:
            if char not in text_freq_map:
                return 0

            text_multiple = text_freq_map[char] // balloon_freq_map[char]

            min_count = min(min_count, text_multiple)

        return min_count


if __name__ == "__main__":
    sol = Solution()

    assert sol.maxNumberOfBalloons("nlaebolko") == 1
    assert sol.maxNumberOfBalloons("loonbalxballpoon") == 2
    assert sol.maxNumberOfBalloons("leetcode") == 0
    assert sol.maxNumberOfBalloons("balloonballoon") == 2
    assert sol.maxNumberOfBalloons("balloonballoooon") == 2
    assert sol.maxNumberOfBalloons("bbaall") == 0

    print("All test cases passed.")
